January 2025 Diffcalc/PP release

stanriders opened the pull request:
https://github.com/ppy/osu-infrastructure/issues/31

This PR includes all changes listed in [the diffcalc project](https://github.com/orgs/ppy/projects/12/views/1) marked as `Ready for deploy`. Some changes were already merged to `master` before `pp-dev` branch was created

Deployment considerations:

- Includes 3 rulesets: **osu**, **taiko** and **catch**. Mania only has a small refactor change that doesn't affect values
- All rulesets have difficulty changes, catch doesn't have performance changes
- **1** new difficulty attribute for osu: `AimDifficultSliderCount`
- **5** difficulty attributes have been removed: `ApproachRate`, `OverallDifficulty`, `GreatHitWindow`, `OkHitWindow`, `ScoreMultiplier`
- New taiko attributes are **NOT** to be databased

osu! changes: https://pp.huismetbenen.nl/rankings/players/master
Taiko changes: https://pp.huismetbenen.nl/rankings/players/master_taiko
Newspost: https://github.com/ppy/osu-wiki/pull/12699

Lawtrohux committed:
Implement considerations for Relax within osu!taiko diffcalc (#30591)

YaniFR committed:
Adjust `DifficultyValue` curve to avoid lower star rating of osu!taiko being too inflated (#31067)

* low sr

* merge two line

* update decimal

* fix formatting

---------

Co-authored-by: StanR <hi@stanr.info>

Lawtrohux committed:
Refactor `TaikoDifficultyCalculator` and add `DifficultStrain` attributes (#31191)

* refactor + countdifficultstrain

* norm in utils

* adjust scaling shift

* fix comment

* revert all value changes

* add the else back

* remove cds comments

Lawtrohux committed:
Increase `accscalingshift` and include `countok` in hit proportion (#31195)

* revert acc scaling shift to previous values

* increase variance in accuracy values across od

* move return values, move nullcheck into return

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Lawtrohux committed:
remove particular condition on convert nerf (#31196)

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Simplify osu! high-bpm acute angle jumps bonus (#30902)

* Simplify osu! high-bpm acute angle jumps bonus

* Add aim wiggle bonus

* Add hitwindow-based aim velocity decrease

* Revert "Add hitwindow-based aim velocity decrease"

This reverts commit bcebe9662cfcb7a72805e48712525ef54ec9820e.

* Move wiggle multiplier to a const, slightly decrease acute bonus multiplier

* Make sure the previous object in the wiggle bonus is also part of the wiggle

* Scale the wiggle bonus multiplayer down

* Increase the acute angle jump bonus multiplier

* Make wiggle bonus only apply on >150 bpm streams, make repetitive angle penalty

* Reduce wiggle bonus multiplier to not break velocity>difficulty relation

* Adjust wiggle falloff function to fix stability issues

* Adjust wiggle consts

* Update tests

Lawtrohux committed:
Implement `Reading` Skill into osu!taiko (#31208)

stanriders committed:
Change slider drop penalty to use actual number of difficult sliders, fix slider drop penalty being too lenient (#31055)

* Change slider drop penalty to use actual number of difficult sliders, fix slider nerf being too lenient

* Move cubing to performance calculation

* Add separate list for slider strains

* Rename difficulty atttribute

* Rename attribute in perfcalc

* Check if AimDifficultSliderCount is more than 0, code quality fixes

* Add `AimDifficultSliderCount` to the list of databased attributes

* Code quality

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

tsunyoku committed:
Use `lastAngle` when nerfing repeated angles on acute bonus (#31245)

* Use `lastAngle` when nerfing repeated angles on acute bonus

* Bump acute multiplier

* Correct outdated wiggle bonus comment

* Update test

---------

Co-authored-by: StanR <hi@stanr.info>

Lawtrohux committed:
Rewrite of the `Rhythm` Skill within osu!taiko (#31284)

* implement bell curve into diffcalcutils

* remove unneeded attributes

* implement new rhythm skill

* change dho variables

* update dho rhythm

* interval interface

* implement rhythmevaluator

* evenhitobjects

* evenpatterns

* evenrhythm

* change attribute ordering

* initial balancing

* change naming to Same instead of Even

* remove attribute bump for display

* Fix diffcalc tests

---------

Co-authored-by: StanR <hi@stanr.info>

tsunyoku committed:
Add basic difficulty & performance calculation for Autopilot mod on osu! ruleset (#21211)

* Set speed distance to 0

* Reduce speed & flashlight, remove aim

* Remove speed AR bonus

* cleanup autopilot mod check in `SpeedEvaluator`

* further decrease speed rating for extra hand availability

* Pass all mods to the speed evaluator, zero out distance bonus instead of distance

---------

Co-authored-by: tsunyoku <mbruhyo@gmail.com>
Co-authored-by: StanR <hi@stanr.info>

stanriders committed:
Fix angle bonuses calculating repetition incorrectly, apply distance scaling to wide bonus (#31320)

* Fix angle bonuses calculating repetition incorrectly, apply distance scaling to wide bonus

* Buff speed to compensate for streams losing pp

* Adjust speed multiplier

* Adjust wide scaling

* Fix tests

Lawtrohux committed:
Add `consistentRatioPenalty` to the `Colour` skill. (#31285)

* fix colour

* review fix

Co-authored-by: StanR <castl@inbox.ru>

* remove cancelled out operand

* increase nerf, adjust tests

* fix automated spacing issues

* up penalty

* adjust tests

* apply review changes

* fix nullable hell

---------

Co-authored-by: StanR <castl@inbox.ru>

stanriders committed:
Clamp OD in performance calculation to fix negative OD gaining pp (#31447)

Co-authored-by: James Wilson <tsunyoku@gmail.com>

stanriders committed:
Simplify angle bonus formula (#31449)

* Simplify angle bonus formula

* Simplify further

* Simplify acute too

* Tests

stanriders committed:
Apply a bunch of balancing changes to aim (#31456)

* Apply a bunch of balancing changes to aim

* Update tests

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

Givikap120 committed:
Punish speed PP for scores with high deviation (#30907)

stanriders committed:
Use total deviation to scale accuracy on aim, general aim buff (#31498)

* Make aim accuracy scaling harsher

* Use deviation-based scaling

* Bring the balancing multiplier down

* Adjust multipliers, fix incorrect deviation when using slider accuracy

* Adjust multipliers

* Update osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceAttributes.cs

Co-authored-by: James Wilson <tsunyoku@gmail.com>

* Change high speed deviation threshold to 22-27 instead of 20-24

* Update tests

---------

Co-authored-by: James Wilson <tsunyoku@gmail.com>

tsunyoku committed:
Remove problematic total deviation scaling, rebalance aim (#31515)

* Remove problematic total deviation scaling, rebalance aim

* Fix tests

Natelytle committed:
Remove lower cap on deviation without misses (#31499)

Lawtrohux committed:
reading mono nerf (#31510)

molneya committed:
fix spinners not increasing cumulative strain time (#31525)

Co-authored-by: StanR <hi@stanr.info>

Lawtrohux committed:
osu!taiko stamina balancing (#31337)

* stamina considerations

* remove consecutive note count

* adjust multiplier

* add back comment

* adjust tests

* adjusts tests post merge

* use diffcalcutils

---------

Co-authored-by: StanR <hi@stanr.info>

bdach committed:
Bump NVika tool to 4.0.0

Code quality CI runs have suddenly started failing out of nowhere:

- Passing run: https://github.com/ppy/osu/actions/runs/12806242929/job/35704267944#step:10:1
- Failing run: https://github.com/ppy/osu/actions/runs/12807108792/job/35707131634#step:10:1

In classic github fashion, they began rolling out another runner change
wherein `ubuntu-latest` has started meaning `ubuntu-24.04` rather than
`ubuntu-22.04`. `ubuntu-24.04` no longer has .NET 6 bundled.

Therefore, upgrade NVika to 4.0.0 because that version is compatible
with .NET 8.

Lawtrohux committed:
osu!taiko star rating and performance points rebalance (#31338)

* rebalance

* revert pp scaling change

* further rebalancing

* comment

* adjust tests

Lawtrohux committed:
osu!taiko further considerations for rhythm (#31339)

* further considerations for rhythm

* new rhythm balancing

* fix license header

* use isNormal to validate ratio

* adjust tests

---------

Co-authored-by: StanR <hi@stanr.info>

Natelytle committed:
Move error function from osu.Game.Utils to osu.Game.Rulesets.Difficulty.Utils (#31520)

* Move error function implementation to osu.Game.Rulesets.Difficulty.Utils

* Rename ErrorFunction.cs to DifficultyCalculationUtils_ErrorFunction.cs

buyaspacecube committed:
Penalise the reading difficulty of high velocity notes using "note density" (#31512)

* Penalise reading difficulty of high velocity notes at high densities

* Use System for math functions

* Lawtrohux changes

* Clean up density penalty comment

* Swap midVelocity and highVelocity back around

* code quality pass

---------

Co-authored-by: Jay Lawton <jaime@merkids.com>
Co-authored-by: StanR <hi@stanr.info>

bastoo0 committed:
Fix osu!catch "buzz slider" SR abuse (#31126)

* Implement fix for catch buzz sliders SR abuse

* Run formatting

---------

Co-authored-by: StanR <hi@stanr.info>

tsunyoku committed:
Remove redundant angle check (#31566)

Lawtrohux committed:
Improve convert considerations in osu!taiko (#31546)

* return a higher finger count

* implement isConvert

* diffcalc cleanup

* harshen monostaminafactor accuracy curve

* readd comment

* adjusts tests

Lawtrohux committed:
Rebalance stamina length bonus in osu!taiko (#31556)

* adjust straincount to assume 1300

* remove comment

---------

Co-authored-by: StanR <hi@stanr.info>

buyaspacecube committed:
osu!taiko new rhythm penalty for long intervals using stamina difficulty (#31573)

* Replace long interval nerf with a new one that uses stamina difficulty

* Turn tabs into spaces

* Update unit tests

---------

Co-authored-by: StanR <hi@stanr.info>

stanriders committed:
Replace indexed skill access with `skills.OfType<...>().Single()` (#30034)

* Replace indexed skill access with `skills.First(s is ...)`

* Fix comment

* Further refactoring to remove casts

---------

Co-authored-by: Dan Balasescu <smoogipoo@smgi.me>
Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

tsunyoku committed:
Use correct `HitWindows` class for osu!taiko hit windows in difficulty calculator (#31579)

* Use correct `HitWindows` class for osu!taiko hit windows in difficulty calculator

* Remove redundant (and incorrect) hit window creation

* Balance rhythm against hit window changes

stanriders commented:
!diffcalc
RULESET=osu

stanriders commented:
!diffcalc
RULESET=taiko

stanriders commented:
!diffcalc
RULESET=catch

peppy committed:
Merge branch 'master' into pp-dev

peppy commented on osu.Game/Rulesets/Difficulty/DifficultyAttributes.cs:
```diff
@@ -30,6 +30,8 @@ public class DifficultyAttributes
         protected const int ATTRIB_ID_AIM_DIFFICULT_STRAIN_COUNT = 25;
         protected const int ATTRIB_ID_OK_HIT_WINDOW = 27;
         protected const int ATTRIB_ID_MONO_STAMINA_FACTOR = 29;
+        protected const int ATTRIB_ID_AIM_DIFFICULT_SLIDER_COUNT = 31;
```
Are these new attributes documented somewhere? Which ruleset are they for? Should they be persisted to the database (apparently some shouldn't now?)

stanriders replied:
This is one of the osu's new attributes, I'm not sure how we should document them code wise other than just through the usage

peppy commented on osu.Game.Rulesets.Taiko/Difficulty/TaikoPerformanceCalculator.cs:
```diff
@@ -73,7 +73,10 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
         {
-            double difficultyValue = Math.Pow(5 * Math.Max(1.0, attributes.StarRating / 0.115) - 4.0, 2.25) / 1150.0;
+            double baseDifficulty = 5 * Math.Max(1.0, attributes.StarRating / 0.110) - 4.0;
+            double difficultyValue = Math.Min(Math.Pow(baseDifficulty, 3) / 69052.51, Math.Pow(baseDifficulty, 2.25) / 1250.0);
```
These inline undocumented random numbers are something.

Could we just add a single liner explaining where they come from?

Lawtrohux replied:
#31067

![image](https://github.com/user-attachments/assets/cbe5b4f4-41ad-47aa-b9f3-aa44dda7c190)

@YaniFR could you link a desmos link for the modified curve here please?

YaniFR replied:
https://www.desmos.com/calculator/ylhjkb4cww here

peppy commented on osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/IHasInterval.cs:
```diff
@@ -0,0 +1,13 @@
+    /// </summary>
+    public interface IHasInterval
+    {
+        double Interval { get; }
```
What's an interval?

Lawtrohux replied:
Interval referring to the changes in rhythm (a change in interval // spacing of an object), and being outputted eventually as a ratio. Originally having a tiny class for solely an interface was questionable, however it was the best outcome I could think of at the time.

peppy commented on osu.Game.Rulesets.Taiko/Difficulty/Preprocessing/Rhythm/Data/SamePatterns.cs:
```diff
@@ -0,0 +1,55 @@
+    /// <summary>
+    /// Represents <see cref="SameRhythmHitObjects"/> grouped by their <see cref="SameRhythmHitObjects.StartTime"/>'s interval.
+    /// </summary>
+    public class SamePatterns : SameRhythm<SameRhythmHitObjects>
```
this is oop hell. without reading any implementation, none of these class names sound sane.

Lawtrohux replied:
Originally, this was labelled as Even instead of Same, with both referring to the objects being grouped by their ratio (rhythm), thus being a same rhythm object, or a same rhythm pattern.

peppy committed:
Tidy up code quality of `RhythmEvaluator`

peppy reviewed (commented):
initial review. taiko code is weird.

smoogipoo commented on osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:
```diff
@@ -129,13 +135,40 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+                {
+                    // We add tick misses here since they too mean that the player didn't follow the slider properly
+                    // We however aren't adding misses here because missing slider heads has a harsh penalty by itself and doesn't mean that the rest of the slider wasn't followed properly
+                    estimateImproperlyFollowedDifficultSliders = Math.Min(countSliderEndsDropped + countSliderTickMiss, attributes.AimDifficultSliderCount);
```
In the previous code, this was a `Math.Clamp()`. This was brought up in a PR at the time (https://github.com/ppy/osu/pull/30544) so I'd like that to still be the case here.

smoogipoo reviewed (commented):
Initial review (catch and osu!)

tsunyoku commented on osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs:
```diff
@@ -107,6 +118,9 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
                 SpeedDifficultStrainCount = speedDifficultyStrainCount,
                 ApproachRate = preempt > 1200 ? (1800 - preempt) / 120 : (1200 - preempt) / 150 + 5,
                 OverallDifficulty = (80 - hitWindowGreat) / 6,
+                GreatHitWindow = hitWindowGreat,
+                OkHitWindow = hitWindowOk,
+                MehHitWindow = hitWindowMeh,
```
What would you want to do in it's place? Databasing OD and calculating them in the performance calculator?

tsunyoku committed:
Remove `EffectiveBPMPreprocessor`

tsunyoku committed:
Refactor `ColourEvaluator`

tsunyoku committed:
Format `ReadingEvaluator`

tsunyoku committed:
Refactor `StaminaEvaluator`

tsunyoku committed:
Add xmldoc to explain `IHasInterval.Interval`

tsunyoku committed:
Rename rhythm preprocessing objects to be clearer with intent

tsunyoku committed:
Make `rescale` a static method

tsunyoku committed:
Fix typo in `ColourEvaluator`

tsunyoku committed:
Revert `Precision.DefinitelyBigger` usage

tsunyoku committed:
Replace weird `IntervalGroupedHitObjects` inheritance layer

tsunyoku commented:
Also, me and @stanriders cannot address any review comments directly as we can't push to the `pp-dev` branch

smoogipoo commented:
We can't fix that with branch rules. I think a separate PR to `pp-dev` is fine? We can merge that PR only when everything's reviewed + fixed (peppy and I haven't finished our reviews yet) and maintain the single commit quality?

tsunyoku committed:
Fix incorrect namespace

tsunyoku committed:
Keep old behaviour of `double.PositiveInfinity` being the default for `Interval`

tsunyoku committed:
Remove unused default value

BabySnakes101 reviewed (approved):
Taiko looks good, some code and magic could be clearer but hopefully some refactoring can be done in the near future.

tsunyoku committed:
Fix incorrect `startTime` calculation

tsunyoku committed:
Account for floating point errors

tsunyoku committed:
Rename class, rename children to hit objects and groups, make fields un-settable

tsunyoku committed:
Use correct English

Rian8337 committed:
Remove unnecessary strain sorting in difficult slider count (#31724)

tsunyoku committed:
Clamp `estimateImproperlyFollowedDifficultSliders` between 0 and `attributes.AimDifficultSliderCount` (#31736)

peppy committed:
Rename `Colour` / `Rhythm` related fields and classes

peppy committed:
Simplify `TaikoRhythmData`'s ratio computation

peppy committed:
Remove unused `HitObjectInterval`

peppy committed:
Tidy up xmldoc and remove another unused field

peppy committed:
Initial tidy-up pass on `IntervalGroupingUtils`

peppy committed:
Add better documentation for interval change code

peppy committed:
Merge pull request #31636 from tsunyoku/taiko-cleanup

peppy committed:
Fix inspections

peppy committed:
Merge branch 'master' into pp-dev

tsunyoku committed:
Calculate hit windows in performance calculator instead of databased difficulty attributes (#31735)

* Calculate hit windows in performance calculator instead of databased difficulty attributes

* Apply mods to beatmap difficulty in osu! performance calculator

* Remove `GreatHitWindow` difficulty attribute for osu!mania

* Remove use of approach rate and overall difficulty attributes for osu!

* Remove use of hit window difficulty attributes in osu!taiko

* Remove use of approach rate attribute in osu!catch

* Remove unused attribute IDs

* Code quality

* Fix `computeDeviationUpperBound` being called before `greatHitWindow` is set

peppy committed:
Merge branch 'master' into pp-dev

peppy committed:
Add note about weird taiko iteration

peppy committed:
Rename buzz variable per review

peppy commented:
I've glossed over a lot of details here, but I think this is about as good to go as it's going to get in my eyes. @smoogipoo do you want to do another pass before we get this going?

Also I think we want to try and synchronise a client release of this better this time around, so I am not going to merge for the upcoming lazer release. We will delay one release cycle (or even do an interim release) so server updates can be prepared and we can have everything go out roughly-together. Sound good?

peppy committed:
Merge branch 'master' into pp-dev

smoogipoo reviewed (dismissed):
Have done a rough pass and everything seems okay.

bdach committed:
Fix failed scores with no hits on beatmaps with ridiculous mod combinations showing hundreds of pp points awarded (#31741)

See
https://discord.com/channels/188630481301012481/1097318920991559880/1334716356582572074.

On `master` this is actually worse and shows thousands of pp points, so
I guess `pp-dev` is a comparable improvement, but still flagrantly
wrong. The reason why `pp-dev` is better is the `speedDeviation == null`
guard at the start of `computeSpeedValue()` which turns off the rest of
the calculation, therefore not exposing the bug where
`relevantTotalDiff` can go negative. I still guarded it in this commit
just for safety's sake given it is clear it can do very wrong stuff.

peppy committed:
Merge branch 'master' into pp-dev

peppy committed:
Merge branch 'master' into pp-dev
